node/actors/bft/src/lib.rs

```diff
@@ -82,36 +96,46 @@ impl Config {
             // This is the infinite loop where the consensus actually runs. The validator waits for either
             // a message from the network or for a timeout, and processes each accordingly.
             loop {
-                let input = pipe
-                    .recv
-                    .recv(&ctx.with_deadline(replica.timeout_deadline))
-                    .await
-                    .ok();
+                let deadline = *deadline_recv.borrow();
+                let input = pipe.recv.recv(&ctx.with_deadline(deadline)).await.ok();
 
                 // We check if the context is active before processing the input. If the context is not active,
                 // we stop.
                 if !ctx.is_active() {
                     return Ok(());
                 }
 
+                // Check if an asynchronous error was received.
+                if let Ok(res) = err_recv.try_recv() {
+                    return Err(res.err().unwrap());
+                }
+
                 let Some(InputMessage::Network(req)) = input else {
-                    replica.start_new_view(ctx).await?;
```

Comment:
+100

Reply:
I moved it to the management of the replica's event loop [here](https://github.com/matter-labs/era-consensus/blob/8099e4d912327b8b5e1dbcb654472b697c0e61e7/node/actors/bft/src/replica/state_machine.rs#L82).

(to clarify—it was moved there because `replica.run()` now consumes `self`, or at least should permanently own the mutable reference, so it's not possible to call `replica.start_new_view()` from here anymore. I used the `Option::None` mechanism which was already in place to communicate with the event loop through the queue. This is indeed awkward.)

Do you mean to move the timeout management under `replica`? It would not be logically identical to the previous behavior, where messages to `leader` would invalidate the timeout. Was this a shortcoming of the previous implementation?

Reply:
I went ahead and made the fix, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/3620997e2cdf9fd7c4e926230d261992c68c0a01